Define behaviour of wxDC::GetTextExtent() and GetMultiLineTextExtent() for empty strings

vadz opened the pull request:
This restores the old behaviour of `GetMultiLineTextExtent()` in wxMSW which was accidentally changed [breaking things](https://trac.wxwidgets.org/ticket/18825), but also, possibly more controversially, changes the behaviour of `GetTextExtent()` in wxGTK and, I think (but I didn't test this), wxOSX, where it previously returned non-zero height for empty strings.

I am pretty sure that we need to handle this case in the same way everywhere, and it seems preferable to make `wxGCDC` follow the other DC classes, but please let me know if anybody strongly disagrees and would rather prefer to change all the other classes to return non-zero height in this case too.

vadz committed:
Remove CppUnit boilerplate from wxDC::GetTextExtent() unit tests

Drop the test case class and use CATCH macros.

vadz committed:
Factor out wxTextMeasureBase::GetEmptyLineHeight()

No real changes, just refactor to extract a trivial helper function.

vadz committed:
Ensure that pointers in GetMultiLineTextExtent() are always valid

Do the same thing as in GetTextExtent() to simplify the code a little
and also avoid temporary variables.

No real changes.

vadz committed:
Use CallGetTextExtent() directly in GetMultiLineTextExtent()

There is no need to pass by GetTextExtent() now that we already make
sure width/height are non-null in this function, so just construct the
MeasuringGuard slightly earlier and call CallGetTextExtent() directly.

No real changes.

vadz committed:
Return non-zero height from GetMultiLineTextExtent("") in wxMSW

This restores the previous behaviour inadvertently changed by bfeae1922d
(Minor optimizations in GetMultiLineTextExtent(), 2020-06-10) and makes
it official by documenting it and adding tests checking for it.

It wasn't completely obviously if this was intentional or accidental
before, but at least wxStaticText itself relied on the old behaviour,
and chances are that so did some code outside the library, so make this
part of the API now.

See #18825.

vadz committed:
Display wxSize less confusingly if any tests fail

Using "x" as separator between the components created confusion with
hexadecimal notation, resulting in confusing messages like "0x17==0x0"
that seemed to describe comparison of 2 numbers and not 2 wxSize
objects.

Using "*" is also more consistent with wxRect output format.

vadz committed:
Make wxGCDC::GetTextExtent("") return (0, 0)

This seems more logical and is compatible with wxDC in wxMSW and wxGTK2,
as well as other kinds of DC, e.g. wxPostScriptDC.

It also looks like the current behaviour was unintentional as it
happened only because wxGCDCImpl::DoGetTextExtent() always passed all
non-null parameters to wxGraphicsContext::GetTextExtent(), even if it
didn't need the values for all of them, and thus bypassed the special
case for the empty string which was already present in the latter
function.

Fix this, making DoGetTextExtent() more efficient as a side effect (we
now avoid unnecessary calls to pango_layout_iter_get_baseline() in the
most common case), and also add another test for empty string to
wxGraphicsContext itself, for non-GTK case.

Also document this behaviour and add a test checking for it.

csomor commented:
Very strange I've a special note about measuring the line height on a one-space-string in case the string to be measured has zero length in `wxMacCoreGraphicsContext::GetTextExtent` ... 

```
    // In wxWidgets (MSW-inspired) API it is possible to call GetTextExtent()
    // with an empty string to get just the descent and the leading of the
    // font, so support this (mis)use.
```

so is it really true, that on wxMSW this wasn't the previous behavior, has it just changed rather recently ? I'd rather strongly advise against changing it ... there might be quite a lot of people depending on it

csomor reviewed (approved):
I'm fine with this, Thanks

a-wi commented on src/common/dcgraph.cpp:
```diff
@@ -1236,7 +1236,16 @@ void wxGCDCImpl::DoGetTextExtent( const wxString &str, wxCoord *width, wxCoord *
+                        height ? &h : NULL,
+                        descent ? &d : NULL,
+                        externalLeading ? &e : NULL
+                      );
```
With these constructs VS2019 raises some warnings:
src\common\dcgraph.cpp(1255): warning C4701: potentially uninitialized local variable 'e' used
src\common\dcgraph.cpp(1253): warning C4701: potentially uninitialized local variable 'd' used
src\common\dcgraph.cpp(1251): warning C4701: potentially uninitialized local variable 'h' used
src\common\dcgraph.cpp(1257): warning C4701: potentially uninitialized local variable 'w' used

vadz replied:
Thanks, fixed now.

vadz committed:
Avoid uninitialized variable warnings in debug MSVC build

The compiler is smart enough to determine that the variables are always
initialized when they're actually used when using optimizations, but
gives a bunch of C4701 warnings for them if we don't initialize them in
the non-optimized debug builds.

vadz commented:
@paulcor I'd like to merge this soon to have at least a small chance to find about any fallout from this before 3.1.4, so I'm going to do it tomorrow if you don't express any objections to this wxGTK change. TIA!

paulcor reviewed (approved):
No objections
